データベース設計

Yusuke0107 opened the pull request:
# What
データベース設計
# Why
データベース作成の為

ER図のURL：[https://gyazo.com/728f12708c0f7d95c7d19e1e2550a65c](https://gyazo.com/728f12708c0f7d95c7d19e1e2550a65c)

Yusuke0107 committed:
データベース設計

Akinori0123 commented on README.md:
```diff
@@ -1,24 +1,60 @@
+|-----------------|--------|-------------|
+| nickname        | string | null: false |
+| email           | string | null: false, unique: true |
+| password        | string | null: false |
```
passwordカラムのカラム名をencrypted_passwordにしましょう。

■理由
deviseを用いる場合、デフォルトでパスワードは暗号化された上でencrypted_passwordカラムに保存されるためです。

Akinori0123 commented on README.md:
```diff
@@ -1,24 +1,60 @@
+| last_name       | string | null: false |
+| first_name_kana | string | null: false |
+| last_name_kana  | string | null: false |
+| birth_date      | string | null: false |
```
birth_dateカラムが誕生日カラムであれば、型をdate型にしましょう。

■理由
管理の容易さから、年月日はdate型で扱うことが推奨されているためです。

Akinori0123 commented on README.md:
```diff
@@ -1,24 +1,60 @@
+## Itemsテーブル
+| Column          | Type       | Options     |
+|-----------------|------------|-------------|
+| user_id         | references | foreign_key: true |
```
user_idカラムは外部キーカラムであるため、外部キーカラムの設計の仕方に合わせて以下の通りに修正しましょう。
・カラム名を`user`にする
・オプションにNOTNULL制約を追加する

**※他テーブルの全ての外部キーカラムも同様に修正しましょう**

Akinori0123 commented on README.md:
```diff
@@ -1,24 +1,60 @@
+| Column          | Type       | Options     |
+|-----------------|------------|-------------|
+| user_id         | references | foreign_key: true |
+| image           | string     | null: false |
```
imageカラムを削除しましょう。
今回は画像の保存にactivestorageを使用するため、imageカラムは不要になります。

Akinori0123 commented on README.md:
```diff
@@ -1,24 +1,60 @@
+| user_id         | references | foreign_key: true |
+| image           | string     | null: false |
+| item_name       | string     | null: false |
+| description     | string     | null: false |
```
商品説明カラムは複数行の文章が入ることを考慮して`text`型にしましょう。

Akinori0123 commented on README.md:
```diff
@@ -1,24 +1,60 @@
+| condition       | string     | null: false |
+| shopping_cost   | string     | null: false |
+| shopping_origin | string     | null: false |
+| shopping_days   | string     | null: false |
```
カテゴリー・商品の状態・配送料負担・発送元地域・発送までの日数カラムはActivehashを用いた実装になるため、Activehashのルールに沿ったカラム名と型にしましょう。
activehashにつきましては以下カリキュラムをご参考にいただければと思います。

Lesson1 最終課題補足カリキュラム
カテゴリーの選択を工夫しよう

Akinori0123 commented on README.md:
```diff
@@ -1,24 +1,60 @@
+| shopping_cost   | string     | null: false |
+| shopping_origin | string     | null: false |
+| shopping_days   | string     | null: false |
+| price           | string     | null: false |
```
価格は数値ですので型を数値型にしましょう。

Akinori0123 commented on README.md:
```diff
@@ -1,24 +1,60 @@
+| shopping_days   | string     | null: false |
+| price           | string     | null: false |
+| fee             | string     |             |
+| profit          | string     |             |
```
fee,profitカラムを削除しましょう。
priceカラムに保存されている情報からいつでもfee,profitを割り出せるためです。

Akinori0123 commented on README.md:
```diff
@@ -1,24 +1,60 @@
+| Column         | Type       | Options     |
+|----------------|------------|-------------|
+| user_id        | references | foreign_key: true |
+| item_id        | references | foreign_key: true |
```
情報の管理を容易にするため、「購入記録情報」と「発送先情報」を別々のテーブルに分けましょう。
最終的に「ユーザー情報」「商品情報」「購入記録」「発送先情報」の四つのテーブルになるようにしていただければと思います。
詳しくは最終課題コースの以下カリキュラムをご参考ください。

実装に悩んだときのヒント
データベース設計

Akinori0123 commented on README.md:
```diff
@@ -1,24 +1,60 @@
+- belongs_to :user
+
+
+## Commentsテーブル
```
Commentsテーブルを削除しましょう。
furimaの必須機能の実装要件においてコメント投稿機能は存在しないためです。

Akinori0123 reviewed (commented):
レビュー依頼いただきありがとうございます！
コミット内容確認いたしました。コメントいたしましたのでご確認のほどよろしくお願いいたします。

Yusuke0107 committed:
データベース設計の修正（1）

・passwordカラムのカラム名をencrypted_passwordに変更
・birth_dateカラムの型をdate型に変更
・user_idカラムのカラム名をuserに変更しオプションにNOTNULL制約を追加（他テーブルの外部キーも変更）
・imageカラムの削除
・商品説明カラムをtext型に変更
・Activehashを用いて実装するカラムはREADMEから削除し、新しく追加したモデルに記述
・価格をinteger型に変更
・fee,profitカラムの削除
・「購入記録情報」と「発送先情報」を別々のテーブルに分離
・Commentsテーブルの削除

TakashiKozuma commented:
レビュー依頼ありがとうございます！

いただいたレビュー依頼ですが、修正点の差分が確認できない状況です。
コミットとプッシュのし忘れがないかをご確認いただき、修正点の差分を出して再度ご提出をお願いいたします。

OhyaYouki commented on README.md:
```diff
@@ -1,24 +1,49 @@
+| user            | references | null: false, foreign_key: true |
+| item_name       | string     | null: false |
+| description     | text       | null: false |
+| price           | integer    | null: false |
```
itemsテーブルにカラムが不足しています。
今一度実装してみましょう。

OhyaYouki commented on README.md:
```diff
@@ -1,24 +1,49 @@
+| birth_date         | date   | null: false |
+### Association
+- has_many :items
+
```
usersテーブルにordersテーブルのアソシエーションを追加しましょう。

OhyaYouki commented on README.md:
```diff
@@ -1,24 +1,49 @@
+### Association
+- belongs_to :shipping_address
+- belongs_to :item
+- has_one :shipping_address
```
ordersテーブルにshipping_addressのアソシエーションが複数個存在しています。
不要なコードを削除しましょう。

また、usersテーブルとのアソシエーションを追加しましょう。

OhyaYouki commented on README.md:
```diff
@@ -1,24 +1,49 @@
+| Column         | Type       | Options     |
+|----------------|------------|-------------|
+| postal_code    | string     | null: false |
+| state          | string     | null: false |
```
```ruby
| state          | string     | null: false |
```
こちらは都道府県選択を保存するカラムでお間違いないでしょうか。

その場合、今回のプルダウンはactive_hashを利用して実装しますので以下の修正をお願いいたします。
・カラム名の末は`_id`とする
・データ型は`integer`とする

OhyaYouki commented on README.md:
```diff
@@ -1,24 +1,49 @@
+| city           | string     | null: false |
+| street_address | string     | null: false |
+| building_name  | string     |             |
+| phone_number   | string     | null: false |
```
orderに関する外部キーを追加しましょう。

OhyaYouki reviewed (commented):
レビュー依頼ありがとうございます。

コメントいたしましたので、ご確認をお願いいたします。

コードの修正後、コミット・プッシュを行った上で再度レビュー依頼をお送りください。
ご対応のほど、よろしくお願いいたします。

Yusuke0107 committed:
データベース設計の修正（2）

・usersテーブルにordersテーブルのアソシエーションを追加
・itemsテーブルに、誤って削除したカラムを追加
・ordersテーブルにshopping_addressのアソシエーションが複数個存在していた為
不要なコードを削除 / usersテーブルとのアソシエーションを追加
・shopping_addressesテーブルの都道府県を保存するカラムをactive_hashで記述した内容に変更し末尾に_idを追加、integer型に変更した
・shopping_addressesテーブルにorderに関する外部キーを追加

OhyaYouki commented on README.md:
```diff
@@ -1,24 +1,56 @@
+| birth_date         | date   | null: false |
+### Association
+- has_many :items
+- has_one :order
```
```ruby
has_one :order
```
一人のユーザーは複数個の商品を購入することができるので
`has_many`で実装しましょう。

OhyaYouki commented on README.md:
```diff
@@ -1,24 +1,56 @@
+| category_id        | integer    | null: false |
+| condition_id       | integer    | null: false |
+| shopping_cost_id   | integer    | null: false |
+| shopping_days_id   | integer    | null: false |
```
```ruby
| shopping_days_id   | integer    | null: false |
```
`(単数形)_id`としましょう。

複数形の場合、モデルファイルを作成した際にモデル名は単数形の原則に反してしまうためです。

OhyaYouki reviewed (commented):
レビュー依頼ありがとうございます。

コメントいたしましたので、ご確認をお願いいたします。

また、データベース設計の段階では実際のモデルファイルやマイグレーションファイルを作成する必要はございません。
削除をお願いいたします。

コードの修正後、コミット・プッシュを行った上で再度レビュー依頼をお送りください。
ご対応のほど、よろしくお願いいたします。

Yusuke0107 committed:
データベース設計の修正（3）

・モデルファイル、マイグレーションファイルの削除
・UsersテーブルのOrdersテーブルとのアソシエーションをhas_manyに変更
・Itemsテーブルのshopping_daysカラムを単数形に変更

Akinori0123 reviewed (commented):
レビュー依頼いただきありがとうございます！
前回の修正内容確認いたしました。LGTMです。
マージ後、mainブランチでpullして次の実装に進んでいただければと思います。

引き続き学習頑張っていきましょう！
